packages/squiggle-lang/src/rescript/ReducerProject/ReducerProject.res

```diff
@@ -0,0 +1,396 @@
+    let these = this->getStdLib
+    let ofUser = Continuation.minus(those, these)
+    ofUser->InternalExpressionValue.nameSpaceToTypeScriptBindings
+  }
+
+  let buildProjectAccessors = (this: t): ProjectAccessorsT.t => {
+    continuation: Bindings.emptyBindings,
+    stdLib: getStdLib(this),
+    environment: getEnvironment(this),
+  }
+
+  let doRunWithContinuation = (
+    this: t,
+    sourceId: string,
+    continuation: ProjectItem.T.continuation,
+  ): unit => {
+    let accessors = buildProjectAccessors(this)
+    let newItem = this->getItem(sourceId)->ProjectItem.run(continuation, accessors)
+    Belt.Map.String.set(this["items"], sourceId, newItem)->T.Private.setFieldItems(this, _)
+    setContinuation(this, sourceId, accessors.continuation)
+  }
+
+  type runState = (ProjectItem.T.resultArgumentType, ProjectItem.T.continuation)
+
+  let tryRunWithContinuation = (
+    this: t,
+    sourceId: string,
+    (rPrevResult: ProjectItem.T.resultArgumentType, continuation: ProjectItem.T.continuation),
+  ): (ProjectItem.T.resultArgumentType, ProjectItem.T.continuation) => {
+    switch getResult(this, sourceId) {
+    | Some(result) => (result, getContinuation(this, sourceId)) // already ran
+    | None =>
+      switch rPrevResult {
+      | Error(error) => {
+          setResult(this, sourceId, Error(error))
+          (Error(error), continuation)
+        }
+      | Ok(_prevResult) => {
+          doRunWithContinuation(this, sourceId, continuation)
+          (
+            getResult(this, sourceId)->Belt.Option.getWithDefault(rPrevResult),
+            getContinuation(this, sourceId),
+          )
+        }
+      }
+    }
+  }
+
+  let runAll = (this: t): unit => {
+    let runOrder = getTopologicalSort(this)
+    let initialState = (Ok(InternalExpressionValue.IEvVoid), getStdLib(this))
+    let _finalState = Belt.Array.reduce(runOrder, initialState, (currState, currId) =>
+      tryRunWithContinuation(this, currId, currState)
+    )
+  }
+
+  let run = (this: t, sourceId: string): unit => {
+    let runOrder = getRunOrderFor(this, sourceId)
+    let initialState = (Ok(InternalExpressionValue.IEvVoid), getStdLib(this))
+    let _finalState = Belt.Array.reduce(runOrder, initialState, (currState, currId) =>
+      tryRunWithContinuation(this, currId, currState)
+    )
+  }
+
+  let evaluate = (sourceCode: string) => {
+    let project = createProject()
+    setSource(project, "main", sourceCode)
+    runAll(project)
+    (
+      getResult(project, "main")->Belt.Option.getWithDefault(IEvVoid->Ok),
+      getContinuation(project, "main"),
+    )
+  }
+}
+
+/*
+  PUBLIC FUNCTIONS
+*/
+
+// Create a new this to hold the sources, executables, bindings and other data.
+// The this is a mutable object for use in TypeScript.
+let createProject = (): t => Private.createProject()->T.Private.castFromInternalProject
+
+// Answers the array of existing source ids to enumerate over.
+let getSourceIds = (this: t): array<string> =>
```

Comment:
If all of these methods are for `t`, it could be nice to put them in a module, ``module T``. Then the names could refer to that, like, ``createProject`` -> ``ReducerProject.create``